src/middleware/upload.ml

```diff
@@ -7,209 +7,118 @@
+      Lwt.async (fun () -> let%lwt _ = th in Lwt.return_unit) ;
+      request.specific.upload.stream <- stream ;
+      state request
+  else state request
 
 type part = [
-  | `Files of (string * string) list
+  | `File of string * string
   | `Value of string
 ]
 
-let log =
-  Log.sub_log "dream.upload"
-
 let multipart request =
-
-  let rec upload_parts files fields =
-    match%lwt upload request with
-    | `Wrong_content_type ->
-      log.warning (fun log -> log ~request
-        "Content-Type not 'multipart/form-data'");
-      Lwt.return `Wrong_content_type
-
-    | `File (name, filename) ->
-      let buffer = Buffer.create 4096 in
-      let rec upload_file_parts () =
-        match%lwt upload_file request with
-        | None ->
-          Lwt.return (Buffer.contents buffer)
-        | Some chunk ->
-          Buffer.add_string buffer chunk;
-          upload_file_parts ()
-      in
-      let%lwt content = upload_file_parts () in
-      upload_parts ((name, filename, content)::files) fields
-
-    | `Field (name, value) ->
-      upload_parts files ((name, value)::fields)
-
-    | `Done ->
-      (* Group the files by field name. This also reverses them into the
-         correct order as a side effect. *)
-      let files_by_field = Hashtbl.create 16 in
-      files |> List.iter (fun (name, filename, content) ->
-        let files =
-          match Hashtbl.find_opt files_by_field name with
-          | None -> []
-          | Some files -> files
-        in
-        Hashtbl.replace files_by_field name ((filename, content)::files));
-      let file_parts =
-        Hashtbl.fold
-          (fun name files file_parts -> (name, `Files files)::file_parts)
-          files_by_field
-          []
-      in
-
-      (* Tag the ordinary fields. *)
-      let field_parts =
-        fields |> List.map (fun (name, value) -> name, `Value value) in
-
+  let content_type =
+    let ( let* ) = Option.bind in
+    let* content_type = Dream.header "content-type" request in
+    let* content_type = Some (content_type ^ "\r\n") in
+    Result.to_option (Multipart_form.Content_type.of_string content_type) in
+  match content_type with
+  | None -> Lwt.return `Wrong_content_type
+  | Some content_type ->
+    let body = Lwt_stream.from (fun () -> Dream.read request) in
+    match%lwt Multipart_form_lwt.of_stream_to_list body content_type with
+    | Error (`Msg _err) ->
+      Lwt.return `Wrong_content_type (* XXX(dinosaure): better error? *)
+    | Ok (tree, assoc) ->
+      let open Multipart_form in
+      let tree = flatten tree in
+      let fold acc { Multipart_form.header; body= uid; } =
+        let body = List.assoc uid assoc in
+        match Header.content_disposition header with
+        | None -> acc
+        | Some v ->
+          match Content_disposition.name v,
+                Content_disposition.filename v with
+          | Some name, None -> (name, `Value body) :: acc
```

Comment:
This appears incorrect w.r.t. [RFC 7578](https://tools.ietf.org/html/rfc7578#section-4.2):

```
   The file name isn't mandatory
   for cases where the file name isn't available or is meaningless or
   private; this might result, for example, when selection or drag-and-
   drop is used or when the form data content is streamed directly from
   a device.
```

However, again, I don't know if real HTTP clients actually emit no filenames for files under any conditions, or if no filename always comes only from a non-file field. Do you know this?

Reply:
Depending on what you want. For my perspective, `multipart-form` does not dissociate values (from `<input ...>`) and files (from `<input type="file" ...>`) and manipulate just _parts_. According to the RFC, we don't have any real clues to say that this specific part is a file or a value. Only the `filename` parameter can help us and it's not a mandatory (as you said) parameter.

So I'm not really sure about these lines but they want to go further (than the RFC) on the interpretation of parts.

Reply:
Ok, after doing some tests, I think `` `Value`` should be removed from the API entirely. Even though browsers don't seem to generate parts without `filename` for files, and it is difficult (impossible?) to do with curl even when reading STDIN, it *is* possible to use curl to submit a field that has file content inlined. Then, it won't have a filename.

Regardless, it is easier for the user, probably, not to deal with two different cases `` `File`` and `Value`. Apps can decide which field should be interpreted as file(s) based only on the field name. The presence, absence, or value of `filename` probably should not be trusted at all, in general &mdash; it is probably only good for informative purposes, to identify files back to the user.

If you're willing, remove it in this PR :) It's also fine to leave it for me to do later.

I added an example, [**`w-multipart-dump`**](https://github.com/aantron/dream/tree/master/example/w-multipart-dump#files), for inspecting `multipart/form-data` bodies sent by clients (as an aside, it appears CSRF checking is not being triggered in it &mdash; will look into that after this PR is settled).

Reply:
I also noticed that providing zero files causes at least Chrome and Firefox to send

```
-----------------------------250068058125696789421882203568
Content-Disposition: form-data; name="files"; filename=""
Content-Type: application/octet-stream


-----------------------------250068058125696789421882203568--
```

i.e., what appears to be a file with an empty filename and no body. This looks like a complication. I think it's fine not to address that in this PR. It should probably be emitted like that in the streaming API, with a note in the docs. For the easy-to-use API, it should probably be filtered out, also with a note in the docs that this is being done.

Reply:
What do you think about:
```ocaml
type part = { filename : string option; contents : string; }

val multipart : request -> [ `Ok of (string * part) list | ... ]
```

Reply:
That looks right.

Is there any point in trying to emit the per-part headers? AFAIK they are all untrustworthy under "wild" HTTP conditions, though maybe they are useful in closed systems somehow? It seems they are obscure.

Reply:
Yes, I will try to extend the record with an `headers` fields 👍